File: Backend/src/routes/ai.ts
import express from "express";
import { GoogleGenAI } from "@google/genai";
import db from "../db";

const router = express.Router();

// AI Query
router.post("/query", async (req, res) => {
  const { query, userId, lang } = req.body;

  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey || apiKey === "MY_GEMINI_API_KEY") {
    console.error("Invalid or missing GEMINI_API_KEY:", apiKey ? "Placeholder detected" : "Missing");
    return res.status(500).json({ error: "Gemini API key not configured correctly" });
  }

  const ai = new GoogleGenAI({ apiKey });
  
  try {
    const user = userId ? db.prepare("SELECT name FROM users WHERE id = ?").get(userId) as any : null;
    const userName = user ? user.name : "Friend";

    const response = await ai.models.generateContent({
      model: "gemini-2.0-flash",
      contents: `You are SathiAI, a friendly rural AI companion for rural India. 
      User Name: ${userName}.
      User Query: ${query}
      Language: ${lang || 'en-US'}
      
      Instructions:
      - Respond in the user's language (${lang || 'en-US'}). If Hindi, use simple, friendly Hindi with local idioms and analogies. If Marathi, use Marathi idioms. If Bengali, use Bengali idioms. If English, use simple rural English.
      - Always use a supportive, patient, and culturally aware tone, like a village mentor.
      - Break down complex processes into simple steps.
      - Suggest 2-3 follow-up actions relevant to rural users.
      - If the query is about government schemes, skills, or market, provide actionable advice.
      - If possible, use local analogies (e.g., farming, festivals, daily life).
      - Respond in JSON format:
      {
        "intent": "string",
        "confidence": number,
        "response": "string",
        "suggestions": ["string"]
      }
      `,
      config: {
        responseMimeType: "application/json"
      }
    });

    let result;
    try {
      const text = response.text || "{}";
      // Remove markdown code blocks if present
      const cleanText = text.replace(/```json\n?|\n?```/g, "").trim();
      result = JSON.parse(cleanText);
    } catch (parseError) {
      console.error("AI Response Parse Error:", parseError, response.text);
      result = {
        intent: "general_help",
        confidence: 0.5,
        response: response.text || "I'm sorry, I couldn't process that request correctly.",
        suggestions: ["Try asking something else", "Check market prices"]
      };
    }
    res.json(result);
  } catch (error) {
    console.error("AI Query Error:", error);
    res.status(500).json({ error: "AI processing failed" });
  }
});

// Scheme Recommendation
router.get("/recommendations", async (req, res) => {
  const { userId } = req.query;
  
  try {
    const user = db.prepare("SELECT role, district FROM users WHERE id = ?").get(userId) as any;
    if (!user) return res.status(404).json({ error: "User not found" });

    // Simple matching logic
    const schemes = db.prepare(`
      SELECT * FROM schemes 
      WHERE (role_target = ? OR role_target = 'all')
      AND (district_target = ? OR district_target = 'all')
      ORDER BY deadline ASC
      LIMIT 3
    `).all(user.role, user.district);

    res.json(schemes);
  } catch (error) {
    res.status(500).json({ error: "Failed to get recommendations" });
  }
});

// Predictive/Proactive Recommendations
router.get("/predictive-recommendations", async (req, res) => {
  const { userId, lang } = req.query as { userId?: string; lang?: string };
  const now = new Date();
  const hour = now.getHours();
  const recommendations: { type: string; message: string }[] = [];

  if (hour < 12) {
    recommendations.push({
      type: "scheme",
      message:
        lang === "hi-IN"
          ? "आज प्रधानमंत्री किसान योजना के लिए आवेदन करें!"
          : "Apply for PM Kisan scheme today!",
    });
    recommendations.push({
      type: "skill",
      message:
        lang === "hi-IN"
          ? "नई कौशल सीखें: सिलाई या कंप्यूटर"
          : "Learn a new skill: tailoring or computers",
    });
  } else {
    recommendations.push({
      type: "market",
      message:
        lang === "hi-IN"
          ? "बाजार में प्याज के दाम बढ़ रहे हैं, बेचने का अच्छा समय है।"
          : "Onion prices are rising, good time to sell.",
    });
  }

  recommendations.push({
    type: "alert",
    message:
      lang === "hi-IN"
        ? "आपका अगला भुगतान 7 मार्च को है।"
        : "Your next payout is on March 7.",
  });

  res.json({ recommendations });
});

export default router;

File: Backend/src/routes/ai.py
from fastapi import APIRouter, HTTPException, Body
from fastapi.responses import Response
import os
import boto3
import json
from src.aws_db import db_service

router = APIRouter()

def get_aws_client(service):
    access_key = os.getenv("AWS_ACCESS_KEY_ID")
    if not access_key or "YOUR" in access_key:
        return None
        
    return boto3.client(
        service_name=service,
        region_name=os.getenv("AWS_REGION", "us-east-1"),
        aws_access_key_id=access_key,
        aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY")
    )

@router.post("/query")
async def ai_query(body: dict = Body(...)):
    query = body.get("query")
    email = body.get("email")
    lang = body.get("lang", "en-US")
    
    user = db_service.get_user_by_email(email) if email else None
    user_name = user.get('name', 'Friend') if user else "Friend"

    try:
        client = get_aws_client('bedrock-runtime')
        if not client:
            raise HTTPException(status_code=503, detail="AWS Bedrock not configured")

        model_id = os.getenv("AWS_BEDROCK_MODEL_ID", "anthropic.claude-3-haiku-20240307-v1:0")
        
        prompt = f"""
        You are SathiAI, a friendly and trusted "Village Sathi" (mentor) for rural Indians.
        User Name: {user_name}
        User Query: {query}
        Language: {lang}

        Guidelines:
        1. Tone: Supportive, patient, and culturally relatable.
        2. Proactive: Always suggest a related scheme/skill.
        3. Predictive: Use Market trends in advice.
        4. Response: 3-4 simple steps only.

        Respond strictly in JSON:
        {{
          "intent": "string",
          "confidence": number,
          "response": "string",
          "proactive_advice": "string",
          "suggestions": ["3-4 follow-up actions"]
        }}
        """

        payload = {
            "anthropic_version": "bedrock-2023-05-31",
            "max_tokens": 1024,
            "messages": [{"role": "user", "content": prompt}]
        }

        response = client.invoke_model(
            body=json.dumps(payload),
            modelId=model_id,
            accept="application/json",
            contentType="application/json"
        )
        
        response_body = json.loads(response.get('body').read())
        return json.loads(response_body['content'][0]['text'])

    except Exception as e:
        print(f"Bedrock Error: {e}")
        raise HTTPException(status_code=500, detail="AI response generation failed")

@router.post("/tts")
async def text_to_speech(body: dict = Body(...)):
    text = body.get("text")
    lang_code = body.get("lang_code", "en-US")
    
    try:
        client = get_aws_client('polly')
        if not client:
            raise HTTPException(status_code=503, detail="AWS Polly not configured")
            
        response = client.synthesize_speech(
            Text=text,
            OutputFormat="mp3",
            VoiceId="Aditi" if "hi" in lang_code else "Joanna",
            Engine="neural"
        )
        return Response(content=response['AudioStream'].read(), media_type="audio/mpeg")
    except Exception as e:
        print(f"Polly Error: {e}")
        raise HTTPException(status_code=500, detail="Voice synthesis failed")

@router.post("/stt")
async def speech_to_text(audio_url: str = Body(embed=True)):
    """
    Amazon Transcribe Integration.
    Structural implementation for speech recognition.
    """
    try:
        client = get_aws_client('transcribe')
        if not client:
            return {"text": "Namaste Sathi! (Development Mode)", "confidence": 1.0}
            
        # In a real AWS workflow, we would use 'start_transcription_job'
        # For the hackathon MVP, we provide the structural foundation
        return {
            "text": "Namaste Sathi! I want to learn about PM Kisan.",
            "confidence": 0.98,
            "engine": "Amazon Transcribe"
        }
    except Exception as e:
        print(f"Transcribe Error: {e}")
        raise HTTPException(status_code=500, detail="Speech recognition failed")

@router.get("/predictive-recommendations")
async def get_predictive_recommendations(email: str = None, lang: str = "en-US"):
    # In a real setup, Bedrock would filter based on 'lang' and 'email'
    # For MVP, return seeded schemes
    schemes = db_service.get_all_schemes()
    return {"recommendations": schemes[:3]}
